drivers/led/is31fl3194/is31fl3194.c

```diff
@@ -0,0 +1,235 @@
+
+#define IS31FL3194_INDEX_MAX 2
+
+struct is31fl3194_config {
+	struct is31fl3194_ctx ctx;
+	const struct i2c_dt_spec i2c;
+	uint8_t num_leds;
+	const struct led_info *leds_info;
+	const uint8_t *curr_map_max;
+};
+
+struct is31fl3194_data {
+	struct is31fl3194_regs regs;
+};
+
+int is31fl3194_i2c_read(const struct i2c_dt_spec *i2c_spec, uint8_t reg_addr, uint8_t *value,
+			uint8_t len)
+{
+	return i2c_burst_read_dt(i2c_spec, reg_addr, value, len);
+}
+
+int is31fl3194_i2c_write(const struct i2c_dt_spec *i2c_spec, uint8_t reg_addr, uint8_t *value,
+			 uint8_t len)
+{
+	return i2c_burst_write_dt(i2c_spec, reg_addr, value, len);
+}
+
+int is31fl3194_on(const struct device *dev, uint32_t led)
+{
+	const struct is31fl3194_config *config = dev->config;
+	struct is31fl3194_ctx *ctx = (struct is31fl3194_ctx *)&config->ctx;
+	int res;
+
+	res = is31fl3194_out_en_set(ctx, led, 1);
+	if (res) {
+		return res;
+	}
+
+	return is31fl3194_color_update(ctx);
+}
+
+int is31fl3194_off(const struct device *dev, uint32_t led)
+{
+	const struct is31fl3194_config *config = dev->config;
+	struct is31fl3194_ctx *ctx = (struct is31fl3194_ctx *)&config->ctx;
+	int res;
+
+	res = is31fl3194_out_en_set(ctx, led, 0);
+	if (res) {
+		return res;
+	}
+
+	return is31fl3194_color_update(ctx);
+}
+
+static const struct led_info *is31fl3194_led_to_info(const struct is31fl3194_config *config,
+						     uint32_t led)
+{
+	for (int i = 0; i < config->num_leds; i++) {
+		if (config->leds_info[i].index == led) {
+			return &config->leds_info[i];
+		}
+	}
+
+	return NULL;
+}
+
+static int is31fl3194_get_info(const struct device *dev, uint32_t led, const struct led_info **info)
+{
+	const struct is31fl3194_config *config = dev->config;
+	const struct led_info *leds_info;
+
+	leds_info = is31fl3194_led_to_info(config, led);
+	if (!leds_info) {
+		*info = NULL;
+		return -EINVAL;
+	}
+
+	*info = leds_info;
+
+	return 0;
+}
+
+int is31fl3194_set_brightness(const struct device *dev, uint32_t led, uint8_t value)
+{
+	const struct is31fl3194_config *config = dev->config;
+	struct is31fl3194_ctx *ctx = (struct is31fl3194_ctx *)&config->ctx;
+	struct is31fl3194_outx_cl outx_cl = { .cl = value };
+	int res;
+
+	res = is31fl3194_outx_cl_set(ctx, led, outx_cl);
+	if (res) {
+		return res;
+	}
+
+	return is31fl3194_color_update(ctx);
+}
+
+int is31fl3194_write_channels(const struct device *dev, uint32_t start_channel,
+			      uint32_t num_channels, const uint8_t *buf)
+{
+	int res;
+
+	if ((start_channel + num_channels) > IS31FL3194_INDEX_MAX) {
+		return -EINVAL;
+	}
+
+	for (int i = 0; i < num_channels; i++) {
+		res = is31fl3194_set_brightness(dev, start_channel + i, buf[i]);
+		if (res) {
+			return res;
+		}
+	}
+
+	return 0;
+}
+
+static const struct led_driver_api is31fl3194_api = {
+	/* Mandatory callbacks. */
+	.on = is31fl3194_on,
+	.off = is31fl3194_off,
+	/* Optional callbacks. */
+	.get_info = is31fl3194_get_info,
+	.set_brightness = is31fl3194_set_brightness,
```

Comment:
Should you not implement the set_color function ? It looks like configuring the color is supported by the IS31FL3194 LED driver.

Reply:
I think the reason why I didn't is that `set_color` sets the color of a certain LED whereas this driver has three single color LEDs connected to it.

Reply:
Are there 3 LEDs ? Or a single LED with 3 channels ?

In the former case, a user may need to configure the LED by calling a single function (with a single I2C command) rather than setting the brightness of three different LED which are in fact a single physical one. This makes a difference when running software LED effects.

Note that the write_channels function can also be used to do that, but in a much less intuitive way.

Reply:
According to the datasheet the is31fl3194 is a 3-channel LED driver and both options are possible: 3 single color LEDs, or a RGB / RG+W LED. So maybe you should implement `set_color` as well. It is your call and you can consider this comment resolved either way.